src/types/router.rs

```diff
@@ -0,0 +1,630 @@
+                graph.add_edge(from_index, to_index, edge.cost);
+            }
+
+            info!("[4/4] Finalizing the router setup...");
+            for node in nodes {
+                if !node_indices.contains_key(node.as_node()) {
+                    let index = graph.add_node(node.as_node());
+                    node_indices.insert(node.as_node(), index);
+                }
+            }
+
+            info!("✨Done! Router engine is ready to use.");
+            Router {
+                graph,
+                node_indices,
+                edges,
+            }
+        }
+
+        /// Get the NodeIndex struct for a given node. The NodeIndex
+        /// struct is used to reference things in the graph.
+        pub fn get_node_index(&self, node: &Node) -> Option<NodeIndex> {
+            debug!("Node: {:?}", node);
+            self.node_indices.get(node).cloned()
+        }
+
+        /// Get a node by NodeIndex.
+        pub fn get_node_by_id(&self, index: NodeIndex) -> Option<&Node> {
+            debug!("Node id: {:?}", index);
+            if self.graph.contains_node(index) {
+                Some(self.graph[index])
+            } else {
+                None
+            }
+        }
+
+        /// Return the number of edges in the graph.
+        pub fn get_edge_count(&self) -> usize {
+            debug!("Edge count: {}", self.graph.edge_count());
+            self.graph.edge_count()
+        }
+
+        /// Find the shortest path between two nodes.
+        ///
+        /// The petgraph's Dijkstra algorithm is very identical to the
+        /// a star algorithm, so we can use the same function for both.
+        /// The only difference might be how the heuristic function is
+        /// implemented.
+        ///
+        /// # Arguments
+        /// * `from` - The node to start from.
+        /// * `to` - The node to end at.
+        /// * `algorithm` - The algorithm to use.
+        /// * `heuristic` - The heuristic function to use.
+        ///
+        /// # Returns
+        /// A tuple of the total cost and the path consisting of node
+        /// indices.
+        ///
+        /// An empty path with a total cost of 0.0 returned if no path
+        /// is found.
+        ///
+        /// An empty path with a total cost of -1.0 is returned if
+        /// either the `from` or `to` node is not found.
+        pub fn find_shortest_path(
+            &self,
+            from: &Node,
+            to: &Node,
+            algorithm: Algorithm,
+            heuristic_function: Option<fn(NodeIndex) -> f32>,
+        ) -> (f32, Vec<NodeIndex>) {
+            info!(
+                "Finding shortest path from {:?} to {:?} using algorithm {:?}",
+                from.location, to.location, algorithm
+            );
+            if self.get_node_index(from).is_some() && self.get_node_index(to).is_some() {
+                let from_index = self.get_node_index(from).unwrap();
```

Comment:
Unhandled unwraps

Reply:
This is handled by the if statement (is_some()).